video/decode/vd_lavc.c

```diff
@@ -1255,6 +1273,13 @@ static int receive_frame(struct mp_filter *vd, struct mp_frame *out_frame)
             force_fallback(vd);
         } while (!ctx->avctx);
 
+        // If software fallback is disabled and we get here, all hwdec must
+        // have failed. Return EOF immediately.
+        if (ctx->opts->software_fallback == INT_MAX) {
+            MP_WARN(ctx, "Software decoding fallback is disabled.\n");
+            return AVERROR_EOF;
+        }
+
```

Comment:
No, it's still not right, because force_fallback can succeed (`ctx->avctx` is set) for a different hwdec, and you'd then fail out when it's not doing software decoding.

I think you need to put the `software_fallback` evaluation logic inside `select_and_set_hwdec`, which is the only place that knows it is about to give up on all hwdecs and do the fallback. It can then refuse to do so and return something or set a flag or whatever, which you can then propagate up into `force_fallback` and then check that value in `receive_frame` to return `AVERROR_EOF`.